File: src/main/java/net/minestom/server/command/builder/arguments/minecraft/ArgumentItemStack.java
package net.minestom.server.command.builder.arguments.minecraft;

import net.kyori.adventure.key.Key;
import net.kyori.adventure.nbt.BinaryTag;
import net.minestom.server.MinecraftServer;
import net.minestom.server.adventure.MinestomAdventure;
import net.minestom.server.codec.Result;
import net.minestom.server.codec.Transcoder;
import net.minestom.server.command.ArgumentParserType;
import net.minestom.server.command.CommandSender;
import net.minestom.server.command.builder.arguments.Argument;
import net.minestom.server.command.builder.exception.ArgumentSyntaxException;
import net.minestom.server.component.DataComponent;
import net.minestom.server.component.DataComponentMap;
import net.minestom.server.item.ItemStack;
import net.minestom.server.item.Material;
import net.minestom.server.registry.RegistryTranscoder;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;

/**
 * Argument which can be used to retrieve an {@link ItemStack} from its material and with component data.
 * <p>
 * It is the same type as the one used in the /give command.
 * <p>
 * Example: diamond_sword{display:{Name:"{\"text\":\"Sword of Power\"}"}}
 */
public class ArgumentItemStack extends Argument<ItemStack> {

    public static final int INVALID_NBT = 1;
    public static final int INVALID_MATERIAL = 2;
    public static final int INVALID_COMPONENT = 3;

    public ArgumentItemStack(String id) {
        super(id, true);
    }

    @Override
    public ItemStack parse(CommandSender sender, String input) throws ArgumentSyntaxException {
        var reader = new StringReader(input);

        Key materialKey = reader.readKey();
        final Material material = materialKey == null ? null : Material.fromKey(materialKey);
        if (material == null)
            throw new ArgumentSyntaxException("Material is invalid", input, INVALID_MATERIAL);
        if (!reader.hasMore()) {
            return ItemStack.of(material); // Nothing else, we have our item
        }

        DataComponentMap.PatchBuilder components = DataComponentMap.patchBuilder();

        // Parse the declared components
        if (reader.peek() == '[') {
            reader.consume('[');
            final Transcoder<BinaryTag> coder = new RegistryTranscoder<>(Transcoder.NBT, MinecraftServer.process());
            do {
                final boolean remove = reader.peek() == '!';
                if (remove)
                    reader.consume('!');
                final Key componentId = reader.readKey();
                final DataComponent<?> component = DataComponent.fromKey(componentId);
                if (component == null)
                    throw new ArgumentSyntaxException("Unknown item component", input, INVALID_COMPONENT);
                if (components.has(component))
                    throw new ArgumentSyntaxException("Repeated item component", input, INVALID_COMPONENT);

                if (remove)
                    components.remove(component);
                else {
                    reader.consume('=');
                    final Result<Object> componentValueResult = (Result<Object>) component.decode(coder, reader.readTag());
                    components.set((DataComponent<Object>) component, componentValueResult.orElseThrow());
                }

                if (reader.peek() != ']')
                    reader.consume(',');
            } while (reader.peek() != ']');
            reader.consume(']');
        }

        if (reader.hasMore())
            throw new ArgumentSyntaxException("Unexpected remaining input", input, INVALID_NBT);

        return ItemStack.of(material, components.build());
    }

    @Override
    public ArgumentParserType parser() {
        return ArgumentParserType.ITEM_STACK;
    }

    @Override
    public String toString() {
        return String.format("ItemStack<%s>", getId());
    }

    private static class StringReader {
        private String input;
        private int index = 0;

        public StringReader(String input) {
            this.input = input;
        }

        public boolean hasMore() {
            return index < input.length();
        }

        public char peek() {
            if (!hasMore()) {
                throw new ArgumentSyntaxException("Unexpected end of input", input, INVALID_NBT);
            }

            return input.charAt(index);
        }

        public void consume(char c) {
            char next = peek();
            if (next != c) {
                throw new ArgumentSyntaxException("Expected '" + c + "', got '" + next + "'", input, INVALID_NBT);
            }
            index++;
        }

        public @Nullable Key readKey() {
            char c;
            int start = index;
            while (hasMore() && (c = peek()) != '{' && c != '[' && c != '=' && c != ',' && c != ']') {
                index++;
            }
            String key = input.substring(start, index);
            if (!Key.parseable(key)) return null;
            return Key.key(key);
        }

        public BinaryTag readTag() {
            try {
                StringBuilder remainder = new StringBuilder();
                final BinaryTag result = MinestomAdventure.tagStringIO().asTag(input.substring(index), remainder);
                this.input = remainder.toString();
                this.index = 0;

                return result;
            } catch (IOException e) {
                throw new ArgumentSyntaxException("Invalid NBT", input, INVALID_NBT);
            }
        }
    }
}

File: src/main/java/net/minestom/server/command/builder/arguments/minecraft/ArgumentItemStack.kt
package net.minestom.server.command.builder.arguments.minecraft

import org.jglrxavpok.hephaistos.parser.SNBTParser.parse
import net.minestom.server.command.builder.arguments.number.ArgumentNumber
import java.util.function.BiConsumer
import net.minestom.server.utils.binary.BinaryWriter
import net.minestom.server.command.builder.exception.ArgumentSyntaxException
import java.lang.NumberFormatException
import java.lang.NullPointerException
import net.minestom.server.command.builder.NodeMaker
import net.minestom.server.network.packet.server.play.DeclareCommandsPacket
import java.math.BigDecimal
import net.minestom.server.utils.location.RelativeVec
import net.minestom.server.command.builder.arguments.relative.ArgumentRelativeVec
import net.minestom.server.utils.location.RelativeVec.CoordinateType
import net.minestom.server.coordinate.Vec
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentRegistry
import net.minestom.server.entity.EntityType
import net.minestom.server.item.Enchantment
import net.minestom.server.potion.PotionEffect
import java.time.temporal.TemporalUnit
import net.minestom.server.command.builder.arguments.minecraft.ArgumentTime
import it.unimi.dsi.fastutil.chars.CharList
import it.unimi.dsi.fastutil.chars.CharArrayList
import java.lang.IllegalArgumentException
import net.minestom.server.command.builder.arguments.minecraft.ArgumentUUID
import net.kyori.adventure.text.format.NamedTextColor
import net.minestom.server.command.builder.arguments.minecraft.ArgumentColor
import net.minestom.server.command.builder.arguments.minecraft.ArgumentRange
import net.minestom.server.utils.entity.EntityFinder
import net.minestom.server.command.builder.arguments.minecraft.ArgumentEntity
import java.lang.StringBuilder
import net.minestom.server.entity.GameMode
import net.minestom.server.utils.entity.EntityFinder.EntitySort
import net.minestom.server.command.builder.arguments.minecraft.ArgumentIntRange
import java.lang.IllegalStateException
import org.jglrxavpok.hephaistos.nbt.NBT
import org.jglrxavpok.hephaistos.nbt.NBTException
import net.minestom.server.command.builder.arguments.minecraft.ArgumentNbtTag
import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer
import net.minestom.server.command.builder.arguments.minecraft.ArgumentComponent
import net.minestom.server.item.ItemStack
import net.minestom.server.command.builder.arguments.minecraft.ArgumentItemStack
import net.minestom.server.item.Material
import org.jglrxavpok.hephaistos.nbt.NBTCompound
import net.minestom.server.command.builder.arguments.minecraft.ArgumentBlockState
import net.minestom.server.command.builder.arguments.minecraft.ArgumentNbtCompoundTag
import net.minestom.server.command.builder.arguments.minecraft.ArgumentResourceLocation
import net.minestom.server.command.builder.ArgumentCallback
import net.minestom.server.command.builder.suggestion.SuggestionCallback
import net.minestom.server.command.builder.arguments.Argument.ArgumentMap
import net.minestom.server.command.builder.arguments.Argument.ArgumentFilter
import java.util.function.UnaryOperator
import java.util.Locale
import java.lang.SafeVarargs
import net.minestom.server.command.builder.NodeMaker.ConfiguredNodes
import net.minestom.server.command.builder.arguments.number.ArgumentInteger
import net.minestom.server.command.builder.arguments.number.ArgumentDouble
import net.minestom.server.command.builder.arguments.number.ArgumentFloat
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentEnchantment
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentParticle
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentPotionEffect
import net.minestom.server.command.builder.arguments.minecraft.registry.ArgumentEntityType
import net.minestom.server.command.builder.arguments.minecraft.ArgumentFloatRange
import net.minestom.server.command.builder.arguments.relative.ArgumentRelativeBlockPosition
import net.minestom.server.command.builder.arguments.relative.ArgumentRelativeVec3
import net.minestom.server.command.builder.arguments.relative.ArgumentRelativeVec2
import net.minestom.server.command.builder.arguments.number.ArgumentLong
import net.minestom.server.command.builder.CommandContext
import net.minestom.server.command.builder.parser.ValidSyntaxHolder
import net.minestom.server.command.builder.parser.CommandParser
import net.minestom.server.command.builder.CommandResult
import net.minestom.server.command.builder.CommandDispatcher
import net.minestom.server.MinecraftServer
import net.minestom.server.command.builder.arguments.*
import org.jglrxavpok.hephaistos.parser.SNBTParser
import java.io.StringReader

/**
 * Argument which can be used to retrieve an [ItemStack] from its material and with NBT data.
 *
 *
 * It is the same type as the one used in the /give command.
 *
 *
 * Example: diamond_sword{display:{Name:"{\"text\":\"Sword of Power\"}"}}
 */
class ArgumentItemStack(id: String) : Argument<ItemStack>(id, true) {
    @Throws(ArgumentSyntaxException::class)
    override fun parse(input: String): ItemStack {
        return staticParse(input)
    }

    override fun processNodes(nodeMaker: NodeMaker, executable: Boolean) {
        val argumentNode: DeclareCommandsPacket.Node =
            Argument.Companion.simpleArgumentNode(this, executable, false, false)
        argumentNode.parser = "minecraft:item_stack"
        nodeMaker.addNodes(arrayOf(argumentNode))
    }

    override fun toString(): String {
        return String.format("ItemStack<%s>", id)
    }

    companion object {
        const val NO_MATERIAL = 1
        const val INVALID_NBT = 2
        const val INVALID_MATERIAL = 3

        @Deprecated("use {@link Argument#parse(Argument)}")
        @Throws(ArgumentSyntaxException::class)
        fun staticParse(input: String): ItemStack {
            val nbtIndex = input.indexOf("{")
            if (nbtIndex == 0) throw ArgumentSyntaxException("The item needs a material", input, NO_MATERIAL)
            return if (nbtIndex == -1) {
                // Only material name
                val material = Material.fromNamespaceId(input)
                    ?: throw ArgumentSyntaxException("Material is invalid", input, INVALID_MATERIAL)
                ItemStack.of(material)
            } else {
                // Material plus additional NBT
                val materialName = input.substring(0, nbtIndex)
                val material = Material.fromNamespaceId(materialName)
                    ?: throw ArgumentSyntaxException("Material is invalid", input, INVALID_MATERIAL)
                val sNBT = input.substring(nbtIndex).replace("\\\"", "\"")
                val compound: NBTCompound
                compound = try {
                    SNBTParser(StringReader(sNBT)).parse() as NBTCompound
                } catch (e: NBTException) {
                    throw ArgumentSyntaxException("Item NBT is invalid", input, INVALID_NBT)
                }
                ItemStack.fromNBT(material, compound)
            }
        }
    }
}